google-auth-provider/main.go

```diff
@@ -0,0 +1,189 @@
+		os.Exit(1)
+	}
+
+	cookieSecret, err := base64.StdEncoding.DecodeString(opts.AuthCookieSecret)
+	if err != nil {
+		fmt.Printf("failed to decode cookie secret: %v\n", err)
+		os.Exit(1)
+	}
+
+	legacyOpts := options.NewLegacyOptions()
+	legacyOpts.LegacyProvider.ProviderType = "google"
+	legacyOpts.LegacyProvider.ProviderName = "google"
+	legacyOpts.LegacyProvider.ClientID = opts.ClientID
+	legacyOpts.LegacyProvider.ClientSecret = opts.ClientSecret
+
+	oauthProxyOpts, err := legacyOpts.ToOptions()
+	if err != nil {
+		fmt.Printf("failed to convert legacy options to new options: %v\n", err)
+		os.Exit(1)
+	}
+
+	oauthProxyOpts.Server.BindAddress = ""
+	oauthProxyOpts.MetricsServer.BindAddress = ""
+	oauthProxyOpts.Cookie.Refresh = time.Hour
+	oauthProxyOpts.Cookie.Name = "obot_access_token"
+	oauthProxyOpts.Cookie.Secret = string(cookieSecret)
+	oauthProxyOpts.Cookie.Secure = strings.HasPrefix(opts.ObotServerURL, "https://")
+	oauthProxyOpts.UpstreamServers = options.UpstreamConfig{
+		Upstreams: []options.Upstream{
+			{
+				ID:            "default",
+				URI:           "http://localhost:8080/",
+				Path:          "(.*)",
+				RewriteTarget: "$1",
+			},
+		},
+	}
+	oauthProxyOpts.RawRedirectURL = opts.ObotServerURL + "/oauth2/callback"
+	oauthProxyOpts.ReverseProxy = true
+	if opts.AuthEmailDomains != "" {
+		oauthProxyOpts.EmailDomains = strings.Split(opts.AuthEmailDomains, ",")
+	}
+
+	if err = validation.Validate(oauthProxyOpts); err != nil {
+		fmt.Printf("failed to validate options: %v\n", err)
+		os.Exit(1)
+	}
+
+	oauthProxy, err := oauth2proxy.NewOAuthProxy(oauthProxyOpts, oauth2proxy.NewValidator(oauthProxyOpts.EmailDomains, oauthProxyOpts.AuthenticatedEmailsFile))
+	if err != nil {
+		fmt.Printf("failed to create oauth2 proxy: %v\n", err)
+		os.Exit(1)
+	}
+
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "9999"
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/{$}", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(fmt.Sprintf("http://127.0.0.1:%s", port)))
+	})
+	mux.HandleFunc("/obot-get-state", obotGetState(oauthProxy))
+	mux.HandleFunc("/obot-get-icon-url", obotGetIconURL)
+	mux.HandleFunc("/", oauthProxy.ServeHTTP)
+
+	fmt.Printf("listening on 127.0.0.1:%s\n", port)
+	if err := http.ListenAndServe("127.0.0.1:"+port, mux); !errors.Is(err, http.ErrServerClosed) {
+		fmt.Printf("failed to listen and serve: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+type SerializableRequest struct {
+	Method string              `json:"method"`
+	URL    string              `json:"url"`
+	Header map[string][]string `json:"header"`
+}
+
+type SerializableState struct {
+	ExpiresOn         *time.Time `json:"expiresOn"`
+	AccessToken       string     `json:"accessToken"`
+	PreferredUsername string     `json:"preferredUsername"`
+	User              string     `json:"user"`
+	Email             string     `json:"email"`
+}
+
+func obotGetState(p *oauth2proxy.OAuthProxy) http.HandlerFunc {
```

Comment:
It would be great if this function could be imported from a shared package for all auth providers. This is very likely to be exactly the same for all auth providers.

I would be OK with this change being made when we add the next auth provider.

Reply:
Done